### ISSUE TITLE
JobCtx.Save() will overwrite results if used by a callback

### ISSUE THREAD

**Issue by kalbhor** [kalbhor/Tasqueue#11]
Currently a `JobCtx` is available to both callbacks and handlers. Calling `JobCtx.Save()` twice will overwrite the previous results.

**Comment by kalbhor** [comment]
It would help to have some kind of separation at the `JobCtx` level to distinguish between such multiple callers.

**Comment by knadh** [comment]
Don't commit if the job status indicates completion?

**Comment by kalbhor** [comment]
Hmm, I'm unclear. The issue is there's a common `JobCtx` passed to all callbacks (success, failed, retry, etc) and also used to store results. It'll be overridden each time `JobCtx.Save()` is called.

This is how `JobCtx` looks like right now.

```go
// JobCtx is passed onto handler functions. It allows access to a job's meta information to the handler.
type JobCtx struct {
       // backend store for results
	store Results
	// results just holds the results set by calling Save().
	results []byte
	Meta    Meta
}
```

I plan on adding separate fields for each result (success, failed, retry callbacks). 
Something similar to below : 
```go
// JobCtx is passed onto handler functions. It allows access to a job's meta information to the handler.
type JobCtx struct {
       // backend store for results
	store Results
	// results just holds the results set by calling Save().
	results []byte
	successResults []byte
        failedResults []byte
}
```

**Comment by knadh** [comment]
> It'll be overridden each time JobCtx.Save() is called.

It doesn't have to, right? When `Save()` is called, check the job status internally to see if it's actually mutable or throw an error otherwise (`success, failed, retry`)?

**Comment by knadh** [comment]
I mean, why should a completed or failed job be mutable? If a job is complete, it is not correct to save more results to it at an arbitrary point. Results should only be acceptable while the job is in `processing`.

Multiple results makes sense. Every time a `Save()` is called, it could be appended to a list internally and access can always be on `.Results[]`.

**Comment by kalbhor** [comment]
Ah, I agree that a completed job should no longer be able to write results. The issue here is that there's a single `JobCtx`, which may be called multiple times during a lifecycle.

Consider the following example:
```go 
func simpleHandler(b []byte, ctx JobCtx) error {
  if string(b) == "fail" {
        ctx.Save([]byte("result on failure")
        return fmt.Errorf("the job has failed")
  }
  ctx.Save([]byte("result on success")
   
 return nil 
}
```

 If the job fails, the result saved on the previous line (```result on failure```) will be lost if a failure callback is registered. The same goes for the result (```result on success```), which will be overwritten if a success callback is registered. 

Multiple results make sense, but as a client, given a job UUID, I should be able to query all the results saved - 
1. Result saved by processing callback 
2. Result saved by success callback 
3. Result saved by handler

**Comment by knadh** [comment]
> Multiple results make sense, but as a client, given a job UUID, I should be able to query all the results saved -
> Result saved by processing callback
> Result saved by success callback
> Result saved by handler

"results" only make sense for completed jobs, semantically. For other states, it is useful metadata to describe those states. I think it's good enough to have a simple []list of payloads that are saved and retrieved. Whether someone saves results on `failed` or `processing` can be left to them. In most common usecases, only `success` will see results being saved.

**Comment by kalbhor** [comment]
Thanks for the detailed explanation. This makes sense to me now. The intent of results/data being saved can be left to the client and they must be aware of it while retrieving the same.

**Comment by kalbhor** [comment]
Fixed in #22
